reckon-core/src/main/groovy/org/ajoberstar/reckon/core/InventoryService.java

```diff
@@ -0,0 +1,257 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.apache.commons.lang3.builder.ToStringStyle;
+import org.eclipse.jgit.lib.Constants;
+import org.eclipse.jgit.lib.ObjectId;
+import org.eclipse.jgit.lib.Ref;
+import org.eclipse.jgit.lib.Repository;
+import org.eclipse.jgit.revwalk.RevCommit;
+import org.eclipse.jgit.revwalk.RevWalk;
+import org.eclipse.jgit.revwalk.RevWalkUtils;
+import org.eclipse.jgit.revwalk.filter.RevFilter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class InventoryService {
+  private static final Logger logger = LoggerFactory.getLogger(InventoryService.class);
+
+  private final Repository repo;
+  private final Function<Ref, Optional<ReckonVersion>> tagParser;
+
+  InventoryService(Repository repo) {
+    this(repo, tagName -> Optional.of(tagName.replaceAll("^v", "")));
+  }
+
+  InventoryService(Repository repo, Function<String, Optional<String>> tagSelector) {
+    this.repo = repo;
+    this.tagParser =
+        ref -> {
+          String tagName = Repository.shortenRefName(ref.getName());
+          return tagSelector.apply(tagName).flatMap(ReckonVersion::valueOf);
+        };
+  }
+
+  public Inventory get() {
+    try (RevWalk walk = new RevWalk(repo)) {
+      walk.setRetainBody(false);
+
+      ObjectId headObjectId = repo.getRefDatabase().getRef("HEAD").getObjectId();
+
+      if (headObjectId == null) {
+        logger.debug("No HEAD commit. Presuming repo is empty.");
+        return new Inventory(null, null, null, 0, null, null);
+      }
+
+      logger.debug("Found HEAD commit {}", headObjectId);
+
+      RevCommit headCommit = walk.parseCommit(headObjectId);
+
+      Set<TaggedVersion> taggedVersions = getTaggedVersions(walk);
+
+      logger.debug("Found tagged versions: {}", taggedVersions);
+
+      ReckonVersion currentVersion =
+          findCurrent(headCommit, taggedVersions.stream())
+              .map(TaggedVersion::getVersion)
+              .orElse(null);
+      TaggedVersion baseNormal =
+          findBase(walk, headCommit, taggedVersions.stream().filter(TaggedVersion::isNormal));
+      TaggedVersion baseVersion = findBase(walk, headCommit, taggedVersions.stream());
+
+      int commitsSinceBase = RevWalkUtils.count(walk, headCommit, baseNormal.getCommit());
+
+      Set<TaggedVersion> parallelCandidates =
+          findParallelCandidates(walk, headCommit, taggedVersions);
+
+      Set<RevCommit> taggedCommits =
+          taggedVersions.stream().map(TaggedVersion::getCommit).collect(Collectors.toSet());
+      Set<ReckonVersion> parallelVersions =
+          parallelCandidates
+              .stream()
+              .map(version -> findParallel(walk, headCommit, version, taggedCommits))
+              // TODO Java 9 Optional::stream
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.ajoberstar.com/coding_rules#rule_key=squid%3AS1135)